components/player-view.js

```diff
@@ -1,318 +1,137 @@
-    if (this.isYt && duration === 0) {
-      window.AppData = deepAssign(window.AppData, {error: true});
-    }
+    if ((this.props.player === 'youtube') && duration === 0) this.onError();
 
     // here we store the muted prop before it gets set in the
     // setVolume call so we can restore it afterwards.
     const wasMuted = this.props.muted;
 
     // set initial volume
-    this.setVolume({
-      target: {
-        value: this.props.volume
-      }
+    emitter.emit('set-volume', {
+      value: this.props.volume
     });
 
     // set muted/unmuted (must be called before setVolume below)
     if (wasMuted) {
-      this.mute();
+      emitter.emit('mute')
     } else {
-      this.unmute();
+      emitter.emit('unmute')
     }
 
+    emitter.emit('load', {value: duration});
+  }
 
-    window.AppData = deepAssign(window.AppData, {
-      loaded: true,
-      duration: duration
-    });
-
-    requestAnimationFrame(this.step);
-  },
-  componentWillUpdate: function() {
-    this.isYt = !!~this.props.domain.indexOf('youtube.com');
-  },
-  componentDidUpdate: function(prevProps, prevState) {
+  componentDidUpdate(prevProps, prevState) {
     if (prevProps.src !== this.props.src) {
-      this.attachVideoListeners();
+      clearTimeout(this.loadingTimeout);
+      this.loadingTimeout = setTimeout(() => {
+        deepAssign(window.AppData, {
+          error: true
+        });
+      }, 30000); // 30 second timeout
+
+      if (this.props.player === 'audio') {
+        const canvas = this.refs['audio-vis'];
+        if (canvas) emitter.emit('init', {
+          src: this.props.src,
+          canvas: canvas,
+          onLoaded: this.onLoaded.bind(this),
+          onError: this.onError.bind(this)
+        });
+      } else this.attachVideoListeners();
     }
-  },
-  attachVideoListeners: function() {
-    if (!this.props.src) return;
+  }
 
-    this.loadingTimeout = setTimeout(function() {
-      window.AppData.error = true;
-    }, 30000); // 30 second timeout
+  componentWillUnmount() {
+    clearTimeout(this.loadingTimeout);
+    this.removeVideoListeners();
+  }
 
-    if (this.isYt) {
-      const PLAYING = window.YT.PlayerState.PLAYING;
-      const PAUSED = window.YT.PlayerState.PAUSED;
+  attachVideoListeners() {
+    if (!this.props.src) return;
 
-      ytCtrl.init('video', {
-        onReady: this.onLoaded,
-        onStateChange: (ev) => {
-          if (ev.data === PLAYING && !this.props.playing) this.play();
-          else if (ev.data === PAUSED && this.props.playing) this.pause();
-        },
-        onError: (err) => {
-          const url = `https://youtube.com/watch?v=${this.props.id}`;
-          window.AppData.error = `There was an error loading your video from ${url}`;
-          console.error('Error: ytCtrl: ', err); // eslint-disable-line no-console
-        }
+    if (this.props.player === 'youtube') {
+      emitter.emit('init', {
+        onLoaded: this.onLoaded.bind(this)
```

Comment:
`onLoaded` is the same regardless since it is just the method right here in `player-view` component. However, I need to be able to call it at different times depending on which `ctrl` is being used.
For instance. in `audio-ctrl` I need to load the underlying file before calling `onLoaded` in that file.

With YouTube I need to call it withing the `init` method because that is where we initialize the playerMap objects so that we can control playback in one place for all events(play, pause, etc..)